public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx

```diff
@@ -0,0 +1,356 @@
+import { css } from '@emotion/css';
+import { groupBy, isEmpty, last, sortBy, take, uniq } from 'lodash';
+import React, { useCallback, useMemo, useRef, useState } from 'react';
+import { useForm } from 'react-hook-form';
+
+import {
+  ArrayVector,
+  DataFrame,
+  DataFrameJSON,
+  dateTime,
+  Field as DataFrameField,
+  FieldType,
+  getDisplayProcessor,
+  GrafanaTheme2,
+  SortedVector,
+  TimeRange,
+} from '@grafana/data';
+import { fieldIndexComparer } from '@grafana/data/src/field/fieldComparers';
+import { Stack } from '@grafana/experimental';
+import { MappingType, ThresholdsMode } from '@grafana/schema';
+import { Alert, Button, Field, Icon, Input, Label, TagList, useStyles2, useTheme2 } from '@grafana/ui';
+
+import { stateHistoryApi } from '../../../api/stateHistoryApi';
+import { combineMatcherStrings, labelsMatchMatchers, parseMatchers } from '../../../utils/alertmanager';
+import { HoverCard } from '../../HoverCard';
+
+import { LogRecordViewerByTimestamp } from './LogRecordViewer';
+import { LogTimelineViewer } from './LogTimelineViewer';
+import { extractCommonLabels, Line, LogRecord, omitLabels } from './common';
+
+interface Props {
+  ruleUID: string;
+}
+
+const MAX_TIMELINE_SERIES = 20;
+
+const LokiStateHistory = ({ ruleUID }: Props) => {
+  const styles = useStyles2(getStyles);
+  const [instancesFilter, setInstancesFilter] = useState('');
+  const logsRef = useRef<HTMLDivElement[]>([]);
+
+  const { getValues, setValue, register, handleSubmit } = useForm({ defaultValues: { query: '' } });
+
+  const { useGetRuleHistoryQuery } = stateHistoryApi;
+  const timeRange = useMemo(() => getDefaultTimeRange(), []);
+  const {
+    currentData: stateHistory,
+    isLoading,
+    isError,
+    error,
+  } = useGetRuleHistoryQuery({ ruleUid: ruleUID, from: timeRange.from.unix(), to: timeRange.to.unix() });
+
+  const { dataFrames, historyRecords, commonLabels, totalRecordsCount } = useInstanceHistoryRecords(
+    stateHistory,
+    instancesFilter
+  );
+
+  const frameSubset = useMemo(() => take(dataFrames, MAX_TIMELINE_SERIES), [dataFrames]);
```

Comment:
any reason why you're using lodash's `take` instead of just `slice`?

Reply:
I think some of my personal bias towards using Lodash / more readable code entered in to this (which is basically executing [`slice(0, n)`](https://github.com/lodash/lodash/blob/master/take.js#L25-L30). 

There's definitely a trade-off here since some engineers might be more familiar with the std library than lodash's functions but hopefully the function names are descriptive enough.